ls/src/main.rs

```diff
@@ -0,0 +1,329 @@
+                        } else if flags.sort_size {
+                            result.sort_by_key(sort_by_size);
+                            result.reverse();
+                        } else {
+                            // Sort the directory entries by file name by default
+                            result.sort_by_key(sort_by_name);
+                        }
+
+                        if flags.reverse {
+                            result.reverse();
+                        }
+                    }
+                },
+                Err(err) => {
+                    eprintln!("ls: cannot access '{}': {}", file, err);
+                    exit_code = 1;
+
+                    break;
+                },
+            }
+        }
+
+        if flags.all || flags.no_sort {
+            // Retrieve the current directories information. This must
+            // be canonicalize incase the path is relative
+            let current = path::PathBuf::from(file).canonicalize()?;
+
+            let dot = File::from_name(".".to_string(), current.clone(), flags)?;
+
+            // Retrieve the parent path. Default to the current path if the parent doesn't
+            // exist
+            let parent_path =
+                path::PathBuf::from(dot.path.parent().unwrap_or_else(|| current.as_path()));
+
+            let dot_dot = File::from_name("..".to_string(), parent_path, flags)?;
+
+            result.insert(0, dot);
+            result.insert(1, dot_dot);
+        }
+
+        if !flags.comma_separate && flags.show_list() {
+            match print_list(result, &mut writer, flags) {
+                Ok(_) => {},
+                Err(err) => {
+                    eprintln!("ls: cannot access '{}': {}", file, err);
+                    exit_code = 1;
+                },
+            }
+        } else {
+            match print_default(result, &mut writer, flags) {
+                Ok(_) => {},
+                Err(err) => {
+                    eprintln!("ls: cannot access '{}': {}", file, err);
+                    exit_code = 1;
+                },
+            }
+        }
+    }
+
+    if exit_code != 0 {
+        process::exit(exit_code);
+    }
+
+    Ok(())
+}
+
+/// Prints information about a file in the default format
+fn print_default<W: Write>(files: Vec<File>, writer: &mut W, flags: Flags) -> io::Result<()> {
+    for file in files {
+        let file_name = file.file_name();
+
+        if flags.comma_separate {
+            write!(writer, "{}, ", file_name)?;
+        } else {
+            writeln!(writer, "{}", file_name)?;
+        }
+    }
+    if flags.comma_separate {
+        writeln!(writer)?;
+    }
+
+    Ok(())
+}
+
+struct Column {
+    pub alignment: Alignment,
+    width: *mut usize,
+    pub value: String,
+}
+
+impl Column {
+    pub fn from(value: String, width: *mut usize, alignment: Alignment) -> Self {
+        Column { alignment, width, value }
+    }
+
+    pub fn width(&self) -> usize { unsafe { *self.width } }
```

Comment:
I rewrote the formatting portion for the `-l` option. I did this because previously I was collecting the values twice which didn't seem ideal. This allowed me to simplify the writing portion as well (see line 304). However, doing some made the column width calculation complicated. The simplest way seemed to be store the width's pointer reference in each column. But I understand that you advice against unsafe code. Do you have any suggestions on a better way to do this? If you want, I can revert to how I was formatting the output previously.

Reply:
I think you can get away with just using usize, since the type is `Copy`, no need to keep pointers around, both as reference types or raw pointer types, and in the end.

If the concern is the size, `usize` and a raw pointer type have the same size, and if the concern is speed, the integer types are Copy because it often is as fast or faster to copy it than pass a reference/pointer.

But, if you must have for it to be a reference you can refactor the structure to hold a reference to it. It would be something like this:

```rust
struct Column<'a> {
    pub alignment: Alignment,
    width: &'a mut usize,
    pub value: String,
}
```

And all uses of the `*mut` pointer in the code could just be a `&mut` reference. This way it makes the compiler able to check that the reference is always valid and that it is surely is a exclusive borrow.

Reply:
I strong suggest to first test if it works changing to `usize` first, and if it doesn't works like you would like it to work, try using a reference instead.